src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
+                }
                 else
                 {
-                    // TODO: Invalidate resources for successful unsafe methods? Required by RFC
-                    await _next(context);
+                    // Store response and response body together
+                    context.CachedResponse.Body = context.ResponseCacheStream.BufferedStream.ToArray();
+                    context.Cache.Set(context.StorageVaryKey ?? context.StorageBaseKey, context.CachedResponse, context.CachedResponseValidFor);
                 }
             }
-            finally
+        }
+
+        internal static void OnResponseStarting(ResponseCachingContext context)
+        {
+            if (!context.ResponseStarted)
+            {
+                context.ResponseStarted = true;
+                context.ResponseTime = context.Options.SystemClock.UtcNow;
+
+                FinalizeCachingHeaders(context);
+            }
+        }
+
+        internal static void ShimResponseStream(ResponseCachingContext context)
+        {
+            // TODO: Consider caching large responses on disk and serving them from there.
+
+            // Shim response stream
+            context.OriginalResponseStream = context.HttpContext.Response.Body;
+            context.ResponseCacheStream = new ResponseCacheStream(context.OriginalResponseStream, context.Options.MaximumCachedBodySize);
+            context.HttpContext.Response.Body = context.ResponseCacheStream;
+
+            // Shim IHttpSendFileFeature
+            context.OriginalSendFileFeature = context.HttpContext.Features.Get<IHttpSendFileFeature>();
+            if (context.OriginalSendFileFeature != null)
             {
-                context.RemoveResponseCachingState();
+                context.HttpContext.Features.Set<IHttpSendFileFeature>(new SendFileFeatureWrapper(context.OriginalSendFileFeature, context.ResponseCacheStream));
+            }
+
+            // TODO: Move this temporary interface with endpoint to HttpAbstractions
+            context.HttpContext.AddResponseCachingFeature();
+        }
+
+        internal static void UnshimResponseStream(ResponseCachingContext context)
+        {
+            // Unshim response stream
+            context.HttpContext.Response.Body = context.OriginalResponseStream;
+
+            // Unshim IHttpSendFileFeature
+            context.HttpContext.Features.Set(context.OriginalSendFileFeature);
+
+            // TODO: Move this temporary interface with endpoint to HttpAbstractions
+            context.HttpContext.RemoveResponseCachingFeature();
+        }
+
+        // Normalize order and casing
+        internal static StringValues GetNormalizedStringValues(StringValues stringVales)
+        {
+            if (stringVales.Count == 1)
+            {
+                return new StringValues(stringVales.ToString().ToUpperInvariant());
+            }
+            else
+            {
+                var originalArray = stringVales.ToArray();
```

Comment:
Sort of. Not all `IList` functionalities are implemented. It's more like a readonly list. Why? Do you recommend using the IList.Sort() method? That's probably not really beneficial since the underlying data structure is an array so we'll just operate on that instead.